Framework/Indexing/src/SpectrumNumberTranslator.cpp

```diff
@@ -0,0 +1,92 @@
+#include "MantidIndexing/SpectrumNumberTranslator.h"
+
+namespace Mantid {
+namespace Indexing {
+
+SpectrumNumberTranslator::SpectrumNumberTranslator(
+    const std::vector<SpectrumNumber> &spectrumNumbers,
+    const Partitioner &partitioner, const PartitionIndex &partition)
+    : m_partition(partition) {
+  partitioner.checkValid(m_partition);
+
+  size_t currentIndex = 0;
+  for (size_t i = 0; i < spectrumNumbers.size(); ++i) {
+    auto partition = partitioner.indexOf(GlobalSpectrumIndex(i));
+    auto number = spectrumNumbers[i];
+    m_partitions.emplace(number, partition);
+    if (partition == m_partition) {
+      m_indices.emplace(number, currentIndex);
+      m_globalToLocal.emplace(GlobalSpectrumIndex(i), currentIndex);
+      ++currentIndex;
+    }
+  }
+  if (spectrumNumbers.size() != m_partitions.size())
+    throw std::logic_error("SpectrumNumberTranslator: The vector of spectrum "
+                           "numbers contained duplicate entries.");
+}
+
+// Full set
+SpectrumIndexSet SpectrumNumberTranslator::makeIndexSet() const {
+  return SpectrumIndexSet(m_indices.size());
+}
+
+SpectrumIndexSet
+SpectrumNumberTranslator::makeIndexSet(SpectrumNumber min,
+                                       SpectrumNumber max) const {
+  // Range check
+  static_cast<void>(m_partitions.at(min));
+  static_cast<void>(m_partitions.at(max));
+
+  // The order of spectrum numbers can be arbitrary so we need to iterate.
+  std::vector<size_t> indices;
+  for (const auto &index : m_indices) {
+    if (index.first >= min && index.first <= max)
+      indices.push_back(index.second);
+  }
+  return SpectrumIndexSet(indices, m_indices.size());
```

Comment:
Here for example you supply `indices` and `indices.size()` could that not be retrieved internally? Then you could compare the size of the set you create with the size of indices which would tell you if there were duplicate indices. Or the argument could be a set so the user is unable to provide duplicate indices?